sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala

```diff
@@ -1719,7 +1719,8 @@ class SessionCatalog(
       name: FunctionIdentifier,
       arguments: Seq[Expression],
       registry: FunctionRegistryBase[T],
-      createFunctionBuilder: CatalogFunction => FunctionRegistryBase[T]#FunctionBuilder): T = {
+      createFunctionBuilder: CatalogFunction => FunctionRegistryBase[T]#FunctionBuilder
+  ): T = synchronized {
```

Comment:
which part is not thread-safe? `qualifyIdentifier`?

Reply:
Let's say we have two threads. If a persistent function has not been loaded into the function registry, the `if` condition here can return false for both threads.
```
if (registry.functionExists(qualifiedIdent)) {
  // This function has been already loaded into the function registry.
  registry.lookupFunction(qualifiedIdent, arguments)
} 
```
Then, if one thread successfully registers the persistent function in 
```
registerFunction(
    funcMetadata,
    overrideIfExists = false,
    registry = registry,
    functionBuilder = createFunctionBuilder(funcMetadata))
```
Then another thread can throw exceptions when trying to register this function again, due to `overrideIfExists = false`

Reply:
ah I see! Is this a long-standing bug?

Reply:
and can we put it as a code comment?

Reply:
> Is this a long-standing bug?

Yes, we should backport it